File: retail_desa_api/app/Models/ActivityLog.php
<?php
namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class ActivityLog extends Model
{
    protected $fillable = [
        'user_id',
        'user_name',
        'user_role',
        'action',
        'product_id',
        'product_name',
        'sku',
        'qty_change',
        'old_stock',
        'new_stock',
        'description',
    ];

    protected $casts = [
        'qty_change' => 'integer',
        'old_stock'  => 'integer',
        'new_stock'  => 'integer',
    ];

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function product()
    {
        return $this->belongsTo(Product::class);
    }

    /**
     * Helper to log activity
     */
    public static function log($user, string $action, ?Product $product = null, int $qtyChange = 0, ?int $oldStock = null, ?int $newStock = null, ?string $description = null)
    {
        return static::create([
            'user_id'      => $user?->id,
            'user_name'    => $user?->name ?? 'Sistem',
            'user_role'    => $user?->role ?? 'system',
            'action'       => $action,
            'product_id'   => $product?->id,
            'product_name' => $product?->name,
            'sku'          => $product?->sku,
            'qty_change'   => $qtyChange,
            'old_stock'    => $oldStock,
            'new_stock'    => $newStock,
            'description'  => $description,
        ]);
    }
}

File: lib/models/activity_log.dart
class ActivityLog {
  final int id;
  final int? userId;
  final String userName;
  final String userRole;
  final String action;
  final int? productId;
  final String? productName;
  final String? sku;
  final int qtyChange;
  final int? oldStock;
  final int? newStock;
  final String? description;
  final DateTime createdAt;

  ActivityLog({
    required this.id,
    this.userId,
    required this.userName,
    required this.userRole,
    required this.action,
    this.productId,
    this.productName,
    this.sku,
    required this.qtyChange,
    this.oldStock,
    this.newStock,
    this.description,
    required this.createdAt,
  });

  factory ActivityLog.fromJson(Map<String, dynamic> json) {
    return ActivityLog(
      id: json['id'] as int,
      userId: json['user_id'] as int?,
      userName: json['user_name'] ?? 'Sistem',
      userRole: json['user_role'] ?? 'system',
      action: json['action'] ?? '',
      productId: json['product_id'] as int?,
      productName: json['product_name'] as String?,
      sku: json['sku'] as String?,
      qtyChange: json['qty_change'] as int? ?? 0,
      oldStock: json['old_stock'] as int?,
      newStock: json['new_stock'] as int?,
      description: json['description'] as String?,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
    );
  }

  String get actionLabel {
    switch (action) {
      case 'create_product':
        return 'Tambah Produk Baru';
      case 'add_stock':
        return 'Tambah Stok';
      case 'reduce_stock':
        return 'Kurangi Stok';
      case 'update_product':
        return 'Edit Produk';
      case 'delete_product':
        return 'Hapus Produk';
      case 'sale_stock_out':
        return 'Penjualan (Stok Keluar)';
      default:
        return action;
    }
  }
}
